internal/controller/object/indexes.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimeevent "sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane-contrib/provider-kubernetes/apis/object/v1alpha2"
+)
+
+const (
+	// resourceRefGVKsIndex is an index of all GroupKinds that
+	// are in use by an Object.
+	resourceRefGVKsIndex = "objectsRefsGVKs"
+	// resourceRefsIndex is an index of resourceRefs that are referenced or
+	// managed by an Object.
+	resourceRefsIndex = "objectsRefs"
+)
+
+var (
+	_ client.IndexerFunc = IndexReferencedResourceRefGVKs
+	_ client.IndexerFunc = IndexReferencesResourcesRefs
+)
+
+// IndexReferencedResourceRefGVKs assumes the passed object is an Object. It
+// returns gvk keys for every resource referenced or managed by the Object.
+func IndexReferencedResourceRefGVKs(o client.Object) []string {
+	obj, ok := o.(*v1alpha2.Object)
+	if !ok {
+		return nil // should never happen
+	}
+
+	// Index references.
+	refs := obj.Spec.References
+	keys := make([]string, 0, len(refs))
+	for _, ref := range refs {
+		refAPIVersion, refKind, _, _ := getReferenceInfo(ref)
+		group, version := parseAPIVersion(refAPIVersion)
+		// References are always on control plane, so we don't pass the config name.
```

Comment:
Idea: 
```suggestion
		providerConfig := "" // references are always local
```

easier to read